### PR TITLE
Upgrade to serde 0.9

### PR DESCRIPTION
**Pull request by jrasky**
Rework things, refactor and fix everything to support serde 0.9.

### CODE REVIEW

**Comment by jrasky** [comment]
@astraw take a look

File: Cargo.toml [hunk 1]
```diff
@@ -11,11 +10,11 @@ keywords = ["messagepack", "binary", "encoding", "serialize"]
...
 [dependencies]
-serde = { version = "0.9", default-features = false, features = ["collections"] }
-byteorder = { version = "1.0", default-features = false }
+serde = { version = "~0.9.8", default-features = false, features = ["collections"] }
+byteorder = { version = "~1.0", default-features = false }
 
-[build-dependencies]
-serde_codegen = { version = "0.9" }
+[dev-dependencies]
+serde_derive = { version = "~0.9.8", features = ["unstable"] }
```
Comment (dtolnay) [inline Cargo.toml]: What unstable feature do you need from serde_derive? I think this cfg actually doesn't do anything in 0.9.8.

**Comment by dtolnay** [comment]
You can skip 0.9 and go directly to Serde 1.0! https://github.com/serde-rs/serde/releases/tag/v1.0.0

**Comment by dtolnay** [comment]
The `from_iter` and `from_bytes` signatures should be:

```rust
pub fn from_iter<I, V>(iter: I) -> Result<V>
    where I: Iterator<Item = u8>,
          V: DeserializeOwned;

pub fn from_bytes<'a, V>(bytes: &'a [u8]) -> Result<V>
    where V: Deserialize<'a>;
```

In `from_iter` it requires that V does not try to borrow any data from the input because iterators don't support that. In `from_bytes` it says V is allowed to borrow from the input bytes of lifetime `'a`.

**Comment by jrasky** [comment]
I need to actually pour over the lifetime documentation for serde. I hacked at it until it compiled, but I'm confident that it's not in good shape. Thanks for pointing that out!

**Comment by jrasky** [comment]
Okay, I think things are in decent shape. I'll merge this in and push to crates.io on monday.

**Comment by astraw** [comment]
@jrasky it is working for me with serde 1.0 on nightly (tests run and my code runs both with and without std). With my PR #6, this is true also for stable rust. Looking forward to the release.

**Pull request merged**